### PR TITLE
Fix controlled vocabulary search for apostrophes and quotes (main branch)

### LINKED ISSUE #9680: Controlled vocabulary search or exact match doesn't work if node value contains '
## Describe the bug
The search for matches in a controlled vocabulary doesn't work if the value contains a '
This is the case both for an exact search 

## To Reproduce
1. Submission eg https://demo.dspace.org/workspaceitems/2246/edit with Subject Keywords input => from controlled vocabulary [srsc.xml](https://github.com/DSpace/DSpace/blob/main/dspace/config/controlled-vocabularies/srsc.xml) with node value eg `Children's language`
2. Search for "Children's language" or "Children's" => no results, as compared to eg "Language technology"

## Expected behavior
Controlled vocabulary should allow search & exact match search for values containing a '

### PR DESCRIPTION
**Pull request by ZahraaChreim-Atmire**
## References
* Fixes #9680
* 7.x version: #9697

## Description
Updated the labelTemplate and xpathTemplate in DSpaceControlledVocabulary to correctly handle searches and exact matches for node values containing apostrophes and quotes, implemented logic in getBestMatch and getMatches to appropriately encase these values by either single or double quotes, or using concat for both, and added integration tests to verify functionality across various input cases.

## Instructions for Reviewers
List of changes in this PR:
* Removed single quote encasing from xpathTemplate and labelTemplate.
* Updated getMatches and getBestMatch to use escapeQuotes method for formatting text hierarchy.
* Added a new private method escapeQuotes to determine whether to encase in single quotes (if containing double quotes or containing neither), encase in double quotes (if containing apostrophe), or concatenate in sections if containing both.
* Added test methods in VocabularyEntryDetailsIT to verify search functionality for scenarios with apostrophes, quotes, both apostrophes and quotes, commas, and colons.

How to test and review this PR:
* Steps to reproduce the bug can be found in #9680
* Queries example endpoint:
    - [/server/api/submission/vocabularies/srsc/entries?filter=Children%27s&exact=false](https://demo.dspace.org/server/api/submission/vocabularies/srsc/entries?filter=Children%27s&exact=false) => should return entry with display value of (_Children%27s language_)

    - [/server/api/submission/vocabularies/srsc/entries?filter=Children%27s language&exact=true](https://demo.dspace.org/server/api/submission/vocabularies/srsc/entries?filter=Children%27s%20language&exact=true) => should return entry with display value of (_Children%27s language_)

    - [/server/api/submission/vocabularies/srsc/entries?filter=Human geography, economic&exact=false](https://demo.dspace.org/server/api/submission/vocabularies/srsc/entries?filter=Human%20geography,%20economic&exact=false) => should return entry with display value of (_Human geography, economic geography_)

    - [/server/api/submission/vocabularies/srsc/entries?filter=Human geography, economic geography&exact=true](https://demo.dspace.org/server/api/submission/vocabularies/srsc/entries?filter=Human%20geography,%20economic%20geography&exact=true) => should return entry with display value of (_Human geography, economic geography_)

    - Add a new input node with label=(_University%27s s"P"apers_) to [srsc.xml](https://github.com/DSpace/DSpace/blob/main/dspace/config/controlled-vocabularies/srsc.xml) and test with [/server/api/submission/vocabularies/srsc/entries?filter=University%27s s"P"apers&exact=true](https://demo.dspace.org/server/api/submission/vocabularies/srsc/entries?filter=University%27s%20s"P"apers&exact=true) => should return entry with display value of (_University%27s s\"P\"apers_)

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code.
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace-api/src/test/data/dspaceFolder/config/controlled-vocabularies/srsc.xml
[the remark this answers is not in the document]
Comment (ZahraaChreim-Atmire) [reply dspace-api/src/test/data/dspaceFolder/config/controlled-vocabularies/srsc.xml:1]: @tdonohue Thank you for the feedback. I've removed the unnecessary second copy of `srsc.xml` and verified that the IT tests still pass.

**Review by tdonohue — COMMENTED** [review]
@ZahraaChreim-Atmire : Thanks for this PR!  We reviewed this together in [today's Developer Meeting](https://wiki.lyrasis.org/display/DSPACE/2024-10-31+DSpace+Developers+Meeting).  The overall feedback was positive.  The code looks good overall, and we really appreciate the detailed Integration Tests!

The only additional feedback was:

* Is there any library that has a utility method which will do this apostrophe/quote escaping *for us* (so that we don't need to write it ourselves)?   One suggestion was possibly [`StringEscapeUtils`](https://commons.apache.org/proper/commons-lang/javadocs/api-3.3/org/apache/commons/lang3/StringEscapeUtils.html)   But, I'll admit, I haven't had a chance myself to see if there's an escape method in that utility class that will do exactly what we want.
* We think this needs a basic (manual) test from the User Interface, just to verify that fully fixes the bug (it looks like it should).  I (@tdonohue) can volunteer to do this manual test, if no one else gets to it before me.

Thanks again @ZahraaChreim-Atmire and apologies it's taken us so long to get back to this PR.

**Comment by ZahraaChreim-Atmire** [comment]
> @ZahraaChreim-Atmire : Thanks for this PR! We reviewed this together in [today's Developer Meeting](https://wiki.lyrasis.org/display/DSPACE/2024-10-31+DSpace+Developers+Meeting). The overall feedback was positive. The code looks good overall, and we really appreciate the detailed Integration Tests!
> 
> The only additional feedback was:
> 
> * Is there any library that has a utility method which will do this apostrophe/quote escaping _for us_ (so that we don't need to write it ourselves)?   One suggestion was possibly [`StringEscapeUtils`](https://commons.apache.org/proper/commons-lang/javadocs/api-3.3/org/apache/commons/lang3/StringEscapeUtils.html)   But, I'll admit, I haven't had a chance myself to see if there's an escape method in that utility class that will do exactly what we want.
> * We think this needs a basic (manual) test from the User Interface, just to verify that fully fixes the bug (it looks like it should).  I (@tdonohue) can volunteer to do this manual test, if no one else gets to it before me.
> 
> Thanks again @ZahraaChreim-Atmire and apologies it's taken us so long to get back to this PR.

After testing if the suggested `StringEscapeUtils.ESCAPE_XML10` works the same way as our `escapeQuotes` method, it turned out it doesn’t. It only helps escape special characters, like single and double quotes (`"` to `"` and `'` to `'`), but it doesn’t automatically handle these cases for safe XPath quoting.

Examples:

- `children's` → it becomes `children's` - for the XPath to work, it should be `"children's"`
- `test\"test` → it becomes `test"test` - for the XPath to work, it should be `'test"test'`

Even if we add single or double quotes back to the xpath so it searches `//node[@label='children's']` or `//node[@label="children's"]` and `//node[@label='test"tes']` or `//node[@label="test"tes"]` → It still doesn’t work.

As a conclusion, `StringEscapeUtils` alone does not provide a direct replacement for our `escapeQuotes` method. It is designed for general XML/HTML escaping, not specifically for formatting strings for safe use in XPath expressions. Therefore, our current `escapeQuotes` method is the correct approach for ensuring that strings are safely formatted for XPath, as it directly addresses the requirements of quote handling in XPath expressions.

**Comment by ZahraaChreim-Atmire** [comment]
I have updated the branch with the latest `main` branch to resolve the merge conflicts.
I have also introduced the following changes:
- Since `getBestMatch` now uses `valueTemplate`, there’s no need to use our escapeQuotes because results are based on the “id” in the vocabulary xml and not the “value” of the nodes.
- Updated the `getBestMatch` method to first attempt an exact match based on the id attribute (using `valueTemplate`). If an exact match is found, the result is returned with a confidence level of `Choices.CF_UNCERTAIN`.
- If no exact “id” match is found, the method now falls back to using `getMatches`. This ensures that if the input text is found within any label, a result will still be returned.

**Comment by kanasznagyzoltan** [comment]
[/server/api/submission/vocabularies/srsc/entries?filter=Children%27s&exact=false](https://demo.dspace.org/server/api/submission/vocabularies/srsc/entries?filter=Children%27s&exact=false) => should return entry with display value of (Children%27s language)

image
 

[/server/api/submission/vocabularies/srsc/entries?filter=Children%27s language&exact=true](https://demo.dspace.org/server/api/submission/vocabularies/srsc/entries?filter=Children%27s%20language&exact=true) => should return entry with display value of (Children%27s language)

image


It doesn’t seem to be working correctly with exact=true, might require further testing

**Comment by ZahraaChreim-Atmire** [comment]
> [/server/api/submission/vocabularies/srsc/entries?filter=Children%27s&exact=false](https://demo.dspace.org/server/api/submission/vocabularies/srsc/entries?filter=Children%27s&exact=false) => should return entry with display value of (Children%27s language)
> 
> image
> 
> [/server/api/submission/vocabularies/srsc/entries?filter=Children%27s language&exact=true](https://demo.dspace.org/server/api/submission/vocabularies/srsc/entries?filter=Children%27s%20language&exact=true) => should return entry with display value of (Children%27s language)
> 
> image
> 
> It doesn’t seem to be working correctly with exact=true, might require further testing

Thanks for pointing this out! The issue was caused by quoting in `valueTemplate`. Since `escapeQuotes()` already handles quoting, I updated the template to avoid wrapping `%s` in single quotes and ensured `escapeQuotes()` is called in `getBestMatch`. I also added a test in `VocabularyEntryDetailsIT` to verify the search query. Ready for another review!

**Comment by ZahraaChreim-Atmire** [comment]
I found additional edge cases in `DSpaceControlledVocabularyTest` that need the single quoted `valueTemplate`.
So I re-added the single quotes to `valueTemplate ` and instead added logic to detect quote conflicts and conditionally strip quotes from the template when needed. We now handle all cases more robustly using proper escaping with `concat(...)` when needed. The XPath expression is now always valid regardless of input.

Everything is re-tested and working, ready for another review!

File: dspace-server-webapp/src/test/data/dspaceFolder/config/controlled-vocabularies/srsc.xml [hunk 1]
```diff
@@ -0,0 +1,2266 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<node id="ResearchSubjectCategories" label="Research Subject Categories">
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/data/dspaceFolder/config/controlled-vocabularies/srsc.xml:2]: @ZahraaChreim-Atmire : Maybe I'm missing something, but why do we need to duplicate the entire `srsc.xml` into this test folder?  Shouldn't it already be available to the tests since it exists by default in the `config/controlled-vocabularies/` folder?   Usually we only need to add files to `src/test/data/dspaceFolder/config` if they are **modified** from the default (as that allows us to override the default configs in tests only)

If you  modified this file in some way, then it'd be good to understand how it differs from the default.  It just seems potentially wasteful to duplicate a 2,000 line XML file here.  Another option would be to create a small XML configuration named `test.xml` or similar which can be used only for testing specific scenarios related to controlled vocabularies.

Comment (ZahraaChreim-Atmire) [reply dspace-server-webapp/src/test/data/dspaceFolder/config/controlled-vocabularies/srsc.xml:2]: @tdonohue The reason for including `srsc.xml` under `src/test/data/dspaceFolder/config` is that we needed to extend it with specific nodes to support additional test cases added in `VocabularyEntryDetailsIT`.
These additions primarily target cases such as vocabulary entries containing apostrophes, quotes, colons, etc, which are not present in the default file.
While I considered creating a smaller, test-specific XML, I found that many existing test methods rely on a wide range of standard nodes already present in the default. Duplicating the full file ensures those tests continue to run reliably without unintended side effects.
Please let me know if this addresses your concern or if you'd prefer further adjustments.

**Comment by tdonohue** [comment]
This PR needs to be rebased, and the bug should be re-tested. This same bug looks like it may have been fixed in #11075 (somewhat by accident), as I cannot reproduce #9680 any longer.

**Comment by bram-atmire** [comment]
Closing as the underlying bug (#9680) can no longer be reproduced after PR #11075 was merged. As @tdonohue noted: "This same bug looks like it may have been fixed in #11075 (somewhat by accident), as I cannot reproduce #9680 any longer."

If any edge cases remain, a fresh PR against current main would be cleaner. Thank you @ZahraaChreim-Atmire.
